File: apps/ios/Velix/Core/PushManager.swift
import Foundation
import FirebaseCore
import FirebaseMessaging

/// Registra o token FCM em cada instância logada (ver spec de UX seção 7 —
/// mesmo token físico do aparelho, uma chamada de POST /push/devices por
/// instância, porque cada backend tem sua própria tabela de DeviceToken).
final class PushManager {
    static let shared = PushManager()

    struct RegisterDeviceBody: Encodable {
        let platform = "ios"
        let fcmToken: String
    }

    /// Setado uma vez em `VelixApp` (e nunca retido fortemente daqui) — só
    /// serve pra `handleNotificationTap` decidir qual instância abrir, ver
    /// abaixo.
    weak var instanceStore: InstanceStore?

    var pendingDeepLinkServerId: String?
    var pendingDeepLinkInstanceId: UUID?

    private init() {}

    /// `fcmToken` pode ser `nil` logo após `registerForRemoteNotifications()`
    /// — o Firebase SDK ainda não recebeu o `apnsToken` de volta do
    /// `AppDelegate` ou ainda não terminou de trocar ele pelo token FCM
    /// (troca é assíncrona). Isso é normal, não um erro: só não registra
    /// nada dessa vez.
    func registerCurrentToken(for instance: Instance, apiClient: APIClient) async {
        // ponytail: Messaging.messaging() dá fatalError se FirebaseApp.configure()
        // nunca rodou (sem GoogleService-Info.plist) — mesma guarda do AppDelegate.
        guard FirebaseApp.app() != nil else { return }
        guard let token = Messaging.messaging().fcmToken else { return }
        _ = try? await apiClient.post("/push/devices", body: RegisterDeviceBody(fcmToken: token)) as EmptyResponse
    }

    /// Deep link ao tocar numa notificação.
    ///
    /// Lacuna conhecida (spec de UX, seção 7): o payload do backend só carrega
    /// `serverId`, nunca um identificador de instância. Como o mesmo token
    /// físico do aparelho é registrado em CADA instância logada (ver
    /// `registerCurrentToken`, chamado uma vez por instância), não dá pra
    /// saber pelo token qual instância mandou o push — só o `serverId` do
    /// payload ajudaria, cruzando com a lista de servidores de cada
    /// instância, e essa lista não é cacheada localmente (fora de escopo
    /// desta task).
    ///
    /// Estratégia best-effort adotada:
    /// - Zero ou uma instância logada: sem ambiguidade — a instância ativa
    ///   (`InstanceStore.activeInstance`) é, por definição, a única.
    /// - Duas ou mais instâncias logadas: ambíguo de verdade. Sem uma busca
    ///   cross-instância (exigiria persistir o `/servers` de cada instância
    ///   localmente), a melhor aproximação é assumir que o push veio da
    ///   instância que já está ativa/em primeiro plano. Pode abrir o
    ///   servidor errado se o push vier de uma instância diferente da ativa
    ///   — limitação conhecida, não resolvida aqui.
    func handleNotificationTap(userInfo: [AnyHashable: Any]) {
        guard let serverId = userInfo["serverId"] as? String else { return }
        pendingDeepLinkServerId = serverId
        pendingDeepLinkInstanceId = instanceStore?.activeInstance?.id
    }
}

File: apps/android/app/src/main/java/com/velix/app/core/PushManager.kt
package com.velix.app.core

import android.content.Context
import com.google.firebase.FirebaseApp
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.coroutines.tasks.await
import kotlinx.serialization.Serializable

@Serializable
private data class RegisterDeviceBody(val platform: String = "android", val fcmToken: String)

/** Registra o token FCM em cada instância logada (mesmo token físico do
 * aparelho, um POST /push/devices por instância — espelha o app iOS).
 *
 * Firebase só existe de verdade se `google-services.json` estava presente no
 * build (ver Task 10 Step 1/app/build.gradle.kts) — sem isso, `FirebaseApp`
 * nunca é inicializado e qualquer chamada a `FirebaseMessaging` derrubaria o
 * app. [configure] é chamado uma única vez em `VelixApplication.onCreate()`
 * com um `Context` de verdade e guarda o resultado num Boolean em cache — daí
 * em diante nenhum outro ponto do app precisa segurar uma referência a
 * `Context` só pra checar isso de novo. */
object PushManager {
    private var firebaseConfigured = false

    var pendingDeepLinkServerId: String? = null
    var pendingDeepLinkInstanceId: String? = null

    fun configure(context: Context) {
        firebaseConfigured = FirebaseApp.getApps(context).isNotEmpty()
    }

    /** `token` pode falhar (ou nunca ter sido emitido) mesmo com Firebase
     * configurado — rede, Play Services ausente no emulador, etc. Falha de
     * registro nunca deve derrubar o fluxo de login/onboarding. */
    suspend fun registerCurrentToken(instance: Instance, apiClient: ApiClient) {
        if (!firebaseConfigured) return
        val token = try { FirebaseMessaging.getInstance().token.await() } catch (e: Exception) { return }
        try {
            apiClient.post<Unit>("/push/devices", RegisterDeviceBody(fcmToken = token))
        } catch (e: Exception) {
            // falha de registro não deve derrubar o fluxo de login/onboarding
        }
    }

    fun handleNotificationTap(serverId: String?, activeInstanceId: String?) {
        pendingDeepLinkServerId = serverId
        pendingDeepLinkInstanceId = activeInstanceId
    }
}
